smithy-rules-engine/src/main/java/software/amazon/smithy/rulesengine/language/eval/Value.java

```diff
@@ -0,0 +1,618 @@
+
+        @Override
+        public java.lang.String toString() {
+            return value.toString();
+        }
+
+        public Map<Identifier, Value> getValue() {
+            return this.value;
+        }
+    }
+
+    public static final class Array extends Value {
+        private final List<Value> inner;
+
+        private Array(List<Value> values) {
+            super(SourceLocation.none());
+            this.inner = values;
+        }
+
+        public List<Value> getValues() {
+            return inner;
+        }
+
+        @Override
+        public Type type() {
+            if (inner.isEmpty()) {
+                return Type.array(Type.empty());
+            } else {
+                Type first = inner.get(0).type();
+                if (inner.stream().allMatch(item -> item.type() == first)) {
+                    return Type.array(first);
+                } else {
+                    throw new SourceException("An array cannot contain different types", this);
+                }
+            }
+        }
+
+        @Override
+        public Array expectArray() {
+            return this;
+        }
+
+        public Value get(int idx) {
+            if (this.inner.size() > idx) {
+                return this.inner.get(idx);
+            } else {
+                return new Value.None();
+            }
+        }
+
+        @Override
+        public Node toNode() {
+            return inner.stream()
+                    .map(ToNode::toNode)
+                    .collect(ArrayNode.collect());
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(inner);
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+            Array array = (Array) o;
+            return inner.equals(array.inner);
+        }
+
+        @Override
+        public java.lang.String toString() {
+            StringBuilder sb = new StringBuilder();
+            sb.append("[");
+            sb.append(inner.stream().map(Object::toString).collect(Collectors.joining(", ")));
+            sb.append("]");
+            return sb.toString();
+        }
+    }
+
+    public static final class None extends Value {
+
+        public None() {
+            super(SourceLocation.none());
+        }
+
+        @Override
+        public Type type() {
+            return Type.empty();
+        }
+
+        @Override
+        public boolean isNone() {
+            return true;
+        }
+
+        @Override
+        public Node toNode() {
+            return Node.nullNode();
+        }
+    }
+
+    public static final class Endpoint extends Value {
+        private final java.lang.String url;
+        private final Map<java.lang.String, Value> properties;
+        private final Map<java.lang.String, List<java.lang.String>> headers;
```

Comment:
Are duplicates in this list okay? Should it be a set or does it have other validation?